Add NCS specific changes

simensrostad opened the pull request:
Add NCS specific changes that was previously added on-top of an internal copy of cJSON in sdk-nrf.

**Note**
Upstream cJSON is build with C89 so the changes added in https://github.com/nrfconnect/sdk-cjson/commit/098fc8394057e0f6b55a03fe3c2639e5e6d31294 will most likely not be added upstream due to ```snprintf```.

As for https://github.com/nrfconnect/sdk-cjson/commit/842de7544dbe5203a1803e8aeacc25083c16e28d I don't know if this is something that could be committed upstream or not. The change seems trivial for most applications, but if it is an optimization that accounts for all users of cJSON it makes sense to add it.

AFAIK there is nothing that prohibits https://github.com/nrfconnect/sdk-cjson/commit/b2f80f37542285e8e0d2f2a2b3c90369e3d13faf from being added upstream as it only adds new APIs on pre-existing functionality to cJSON. So I marked it ```toup```.

simensrostad committed:
[nrf noup] Add secure enhancements to cJSON

This patch ensures cJSON does not go of out bounds using `snprintf`
functions.

This patch is marked [nrf noup] due to upstream cJSON enforcing C89
that does not include `snprintf`.

Signed-off-by: Simen S. Røstad <simen.rostad@nordicsemi.no>
Co-authored-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

simensrostad committed:
[nrf fromlist] Be less greedy with heap when reallocating

Change ensure() in cJSON.c to only request what is needed, not double
the size. This is too wasteful when printing large strings.

Upstream PR: https://github.com/DaveGamble/cJSON/pull/577

Signed-off-by: Simen S. Røstad <simen.rostad@nordicsemi.no>
Co-authored-by: Pete Skeggs <peter.skeggs@nordicsemi.no>

simensrostad committed:
[nrf fromlist] Add more CS versions of cJSON_Add functions

Add more constant string key versions of cJSON_Add*
functions to save heap in memory-constrained applications.

Upstream PR: https://github.com/DaveGamble/cJSON/pull/578

Signed-off-by: Simen S. Røstad <simen.rostad@nordicsemi.no>
Co-authored-by: Pete Skeggs <peter.skeggs@nordicsemi.no>

sigvartmh reviewed (approved):
Seems to contain the `strcpy`to`strncpy` and `sprintf` and `snprintf`. So they are now using more safe functions. Thanks Simen :)

mbolivar-nordic commented on cJSON.h:
```diff
@@ -255,7 +266,7 @@ CJSON_PUBLIC(cJSON *) cJSON_Duplicate(const cJSON *item, cJSON_bool recurse);
 
 /* Minify a strings, remove blank characters(such as ' ', '\t', '\r', '\n') from strings.
- * The input pointer json cannot point to a read-only address area, such as a string constant, 
+ * The input pointer json cannot point to a read-only address area, such as a string constant,
```
Please remove the extraneous whitespace change. That seems unrelated to this commit's purpose.

simensrostad replied:
Jepp. Done.

mbolivar-nordic commented on cJSON.c:
```diff
@@ -92,19 +92,19 @@ CJSON_PUBLIC(const char *) cJSON_GetErrorPtr(void)
 }
 
-CJSON_PUBLIC(char *) cJSON_GetStringValue(const cJSON * const item) 
+CJSON_PUBLIC(char *) cJSON_GetStringValue(const cJSON * const item)
```
Please don't mix large-scale whitespace and functional changes like this. It makes things harder to review and adds diff from upstream, which is a source of merge conflicts when synchronizing. Can you go through the commits in this PR and remove the intermixing of whitespace and functional changes within a single commit?

If you're going to fix whitespace nits, I suggest doing it upstream first, since they are not bugs, but style issues, and we want to keep it simple when it comes to avoiding merge conflicts when syncing with upstream later. That means minimizing diff from upstream.

simensrostad replied:
:+1:

simensrostad replied:
I see that I might need to look this one over again.

simensrostad replied:
Should be OK.

mbolivar-nordic commented on cJSON.c:
```diff
@@ -558,22 +558,22 @@ static cJSON_bool print_number(const cJSON * const item, printbuffer * const out
     if (isnan(d) || isinf(d))
     {
-        length = sprintf((char*)number_buffer, "null");
+        length = snprintf((char*)number_buffer, sizeof(number_buffer), "null");
```
What is the justification for having this be a `noup` patch? Did upstream reject it? If so, what is wrong with their justification; if not, can we send this upstream and keep it here as a `fromlist` patch instead?

mbolivar-nordic replied:
Never mind, I see the note in the PR description that explains why upstream probably won't want this.

Can you please put it into the commit log, so we don't lose this information?

simensrostad replied:
Updated the commit log explaining why this patch cannot go upstream.

mbolivar-nordic commented on cJSON.c:
```diff
@@ -479,7 +479,7 @@ static unsigned char* ensure(printbuffer * const p, size_t needed)
     else
     {
-        newsize = needed * 2;
+        newsize = needed + 1;
```
Same question here; any reason not to send this upstream and see what they say, instead of carrying a `noup` patch?

mbolivar-nordic replied:
Likewise I see the note in the PR description now. I still think we should submit it upstream and see what they say.

If upstream has a configuration mechanism for fine-tuning the library behavior, perhaps this can leverage that to make this settable in NCS.

mbolivar-nordic commented on cJSON.c:
```diff
@@ -2057,6 +2057,124 @@ CJSON_PUBLIC(cJSON_bool) cJSON_AddItemToObjectCS(cJSON *object, const char *stri
     return add_item_to_object(object, string, item, &global_hooks, true);
 }
 
+CJSON_PUBLIC(void) cJSON_AddItemReferenceToObjectCS(cJSON *object, const char *string, cJSON *item)
```
We generally try to avoid `toup` patches unless it's an emergency and there really is no time to submit upstream first.

Can you please:

1. submit this patch upstream (without any `[nrf ...]` tag, of course)
2. change this commit to `fromlist`
3. add a link to the upstream PR in the commit log of the downstream `fromlist` patch?

If upstream rejects it, we can carry a `noup`. If upstream is slow to respond, we can carry a `fromlist`. But let's try to get this upstreamed, please.

plskeggs replied:
I missed this partner PR to https://github.com/nrfconnect/sdk-nrf/pull/4409 entirely.  Sorry about that.

Yes, I'll submit my changes upstream when I am less busy with the big push to get predicted GPS into NCS 1.6.0.  Hope more delay is acceptable.

plskeggs replied:
Doh!  I stand corrected -- @simensrostad already did that.  Thanks!

mbolivar-nordic reviewed (changes requested):
Thanks very much for putting this together.

I have a few requests but this looks very good already.

simensrostad commented:
@mbolivar-nordic Added https://github.com/nrfconnect/sdk-cjson/pull/1/commits/7509bd2bda1463a14b3b6a2c6913ecdefb5e862f and https://github.com/nrfconnect/sdk-cjson/pull/1/commits/61407f339b0c1fc7c6374aa3b180c0398d1bb85a upstream and changed the tags to [nrf fromlist]. I also resolved the whitespace issues that you commented on.

tejlmand reviewed (changes requested):
I believe this should should be made into a Zephyr module by adding `zephyr/module.yml`:
```
build:
    cmake-ext: True
    kconfig-ext: True

```

simensrostad committed:
[nrf noup] Make repository a zephyr module

Add module.yml that enables zephyr based projects to include
the repository as a zephyr module.

Signed-off-by: Simen S. Røstad <simen.rostad@nordicsemi.no>

simensrostad commented:
> I believe this should should be made into a Zephyr module by adding `zephyr/module.yml`:
> 
> ```
> build:
>     cmake-ext: True
>     kconfig-ext: True
> ```

Added ```zephyr/module.yml``` to root folder.

plskeggs commented:
LGTM

simensrostad committed:
[nrf noup] Remove CI tests.

Remove CI tests for cJSON.

Signed-off-by: Simen S. Røstad <simen.rostad@nordicsemi.no>

mbolivar-nordic commented:
Sorry about this, but, I had to force push into master after merging this. This repository was not configured for rebase and merge, so clicking 'merge' created a merge commit, which we don't want. I force pushed just these changes. Please update the sdk-nrf PR with the final SHA which was merged into the main branch, 50bdddd.

simensrostad commented:
> Sorry about this, but, I had to force push into master after merging this. This repository was not configured for rebase and merge, so clicking 'merge' created a merge commit, which we don't want. I force pushed just these changes. Please update the sdk-nrf PR with the final SHA which was merged into the main branch, [50bdddd](https://github.com/nrfconnect/sdk-cjson/commit/50bddddf7fb45e4ce87bc3619df06db55d181856).

No worries. The commit hash should now be up to date.

bas-archembedded commented:
@mbolivar-nordic @simensrostad Can you please have a look at https://github.com/nrfconnect/sdk-cjson/pull/3?
